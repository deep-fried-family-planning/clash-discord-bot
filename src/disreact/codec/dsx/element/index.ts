import {S} from '#src/internal/pure/effect.ts';
import * as Function from '#src/disreact/codec/dsx/element/function-element.ts';
import * as Intrinsic from '#src/disreact/codec/dsx/element/intrinsic-element.ts';
import * as Text from '#src/disreact/codec/dsx/element/text-element.ts';
export * as Function from '#src/disreact/codec/dsx/element/function-element.ts';
export * as Intrinsic from '#src/disreact/codec/dsx/element/intrinsic-element.ts';
export * as Text from '#src/disreact/codec/dsx/element/text-element.ts';
export * as Props from 'src/disreact/codec/dsx/common/props.ts';
export * as Children from '#src/disreact/codec/dsx/common/children.ts';
export * as Component from '#src/disreact/codec/dsx/common/component.ts';



export const T = S.Union(
  Function.T,
  Intrinsic.T,
  Text.T,
);

export type T =
  | Function.T
  | Intrinsic.T
  | Text.T;

export const isFunction = Function.is;
export const isIntrinsic = Intrinsic.is;
export const isText = Text.is;
