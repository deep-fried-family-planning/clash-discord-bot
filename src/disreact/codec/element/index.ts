import type { Elem } from '#src/disreact/model/entity/element';
import {RestElement} from '#src/disreact/model/entity/element-rest.ts';
import {TaskElem} from '#src/disreact/model/entity/element-task.ts';
import {LeafElem} from '#src/disreact/model/entity/leaf.ts';



export const cloneElement = (self: Elem): Elem => {
  if (LeafElem.is(self)) {
    return LeafElem.clone(self);
  }

  if (RestElement.isTag(self)) {
    return RestElement.clone(self);
  }

  return TaskElem.clone(self);
};

export const cloneTree = (self: Elem, parent?: Elem): Elem => {
  const linked = linkParent(self, parent);
  const cloned = cloneElement(linked);

  if (cloned.children.length) {
    cloned.children = cloned.children.map((child) => cloneTree(child, cloned));
  }

  return cloned;
};



export const linkParent = <A extends Elem>(node: A, parent?: Elem): A => {
  if (!parent) {
    node.idx     = 0;
    node.id      = `${node.id}:${node.idx}`;
    node.step_id = `${node.id}:${node.idx}`;
    node.full_id = `${node.id}:${node.idx}`;
  }
  else {
    node.id      = `${node.id}:${node.idx}`;
    node.step_id = `${parent.id}:${node.id}`;
    node.full_id = `${parent.full_id}:${node.id}`;
  }
  return node;
};
