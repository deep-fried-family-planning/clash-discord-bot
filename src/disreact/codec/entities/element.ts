import type * as FunctionElement from '#src/disreact/codec/entities/function-element.ts';
import type * as IntrinsicElement from '#src/disreact/codec/entities/intrinsic-element.ts';
import type * as TextElement from '#src/disreact/codec/entities/text-element.ts';


export type Type =
  | TextElement.Type
  | IntrinsicElement.Type
  | FunctionElement.Type;
