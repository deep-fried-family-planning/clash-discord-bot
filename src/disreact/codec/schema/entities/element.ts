import type * as FunctionElement from 'src/disreact/codec/schema/entities/function-element.ts';
import type * as IntrinsicElement from 'src/disreact/codec/schema/entities/intrinsic-element.ts';
import type * as TextElement from 'src/disreact/codec/schema/entities/text-element.ts';


export type Type =
  | TextElement.Type
  | IntrinsicElement.Type
  | FunctionElement.Type;
