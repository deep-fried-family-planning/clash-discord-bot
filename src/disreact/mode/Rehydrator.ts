import * as El from '#src/disreact/mode/entity/el.ts';
import * as FC from '#src/disreact/mode/entity/fc.ts';
import * as Rehydrant from '#src/disreact/mode/entity/rehydrant.ts';
import * as JsxSchema from '#src/disreact/mode/jsx-schema.ts';
import * as Data from 'effect/Data';
import * as E from 'effect/Effect';
import * as Stack from 'effect/MutableList';

const getId = (input: string | Rehydrant.SourceId) => {
  if (typeof input === 'string') return input;
  if (FC.isFC(input)) return input[FC.NameId]!;
  if (El.isComponent(input)) return input.type[FC.NameId]!;
};

export class RehydratorError extends Data.TaggedError('RehydratorError')<{}> {}

export type RehydratorConfig = {
  sources?: | Rehydrant.Registrant[]
            | { [K in string]: Rehydrant.Registrant };

  primitive?    : string;
  normalization?: Record<string, string>;
  encoding?     : Record<string, (self: any, acc: any) => any>;
};

export class Rehydrator extends E.Service<Rehydrator>()('disreact/Rehydrator', {
  effect: (config?: RehydratorConfig) => E.gen(function* () {
    const store         = new Map<string, Rehydrant.Source>(),
          primitive     = config?.primitive ?? JsxSchema.primitive,
          normalization = config?.normalization ?? JsxSchema.normalization,
          encoding      = config?.encoding ?? JsxSchema.encoding,
          sources       = config?.sources ?? [];

    if (Array.isArray(sources)) {
      for (const input of sources) {
        const src = Rehydrant.source(input);
        store.set(src.id, src);
      }
    }
    else if (typeof sources === 'object') {
      for (const [id, input] of Object.entries(sources)) {
        const src = Rehydrant.source(input, id);
        if (store.has(src.id)) {
          return yield* new RehydratorError();
        }
        store.set(src.id, src);
      }
    }

    const register = (input: Rehydrant.Registrant, id?: string) => E.suspend(() => {
      const src = Rehydrant.source(input, id);
      if (store.has(src.id)) {
        return new RehydratorError();
      }
      store.set(src.id, src);
      return E.void;
    });

    const checkout = (input: Rehydrant.SourceId, props: any, data?: any) => E.suspend(() => {
      const id = getId(input);
      if (!id) {
        return new RehydratorError();
      }
      if (!store.has(id)) {
        return new RehydratorError();
      }
      const source = store.get(id)!;
      return E.succeed(Rehydrant.fromSource(source, props, data));
    });

    const decode = (hydrator: Rehydrant.Hydrator, data?: any) => E.suspend(() => {
      if (!store.has(hydrator.id)) {
        return new RehydratorError();
      }
      const source = store.get(hydrator.id)!;
      return E.succeed(Rehydrant.fromHydrator(source, hydrator, data));
    });

    const encode = (root: Rehydrant.Rehydrant | null) => {
      if (!root || El.isText(root.elem)) {
        return null;
      }

      const result = {} as any,
            stack  = Stack.make<[any, El.El[]]>([result, [root.elem]]),
            args   = new WeakMap<El.Nd, any>();

      while (Stack.tail(stack)) {
        const [acc, cs] = Stack.pop(stack)!;
        for (let i = 0; i < cs.length; i++) {
          const c = cs[i];

          if (El.isText(c)) {
            acc[primitive] ??= [];
            acc[primitive].push(c);
          }
          else if (args.has(c) && El.isRest(c)) {
            const norm = normalization[c.type];
            const arg = args.get(c)!;
            acc[norm] ??= [];
            acc[norm].push((encoding[c.type](c, arg)));
          }
          else if (!c.nodes.length && El.isRest(c)) {
            const norm = normalization[c.type];
            const arg = {};
            args.set(c, arg);
            acc[norm] ??= [];
            acc[norm].push((encoding[c.type](c, arg)));
          }
          else {
            Stack.append(stack, [acc, cs.slice(i)]);
            const arg = {};
            args.set(c, arg);
            const next = El.isRest(c) ? arg : acc;
            Stack.append(stack, [next, c.nodes]);
            break;
          }
        }
      }
      for (const key of Object.keys(result)) {
        if (result[key]) {
          return {
            _tag    : key,
            hydrator: Rehydrant.hydrator(root),
            data    : result[key][0],
          };
        }
      }
      return null;
    };

    return {
      register,
      checkout,
      decode,
      encode,
    };
  }),
  accessors: true,
}) {}
