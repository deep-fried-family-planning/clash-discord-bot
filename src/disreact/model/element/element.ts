import {RestElement} from '#src/disreact/model/element/rest.ts';
import {TaskElem} from '#src/disreact/model/element/task.ts';
import {LeafElem} from '#src/disreact/model/element/leaf.ts';
import type {FiberNode} from '#src/disreact/model/entity/fiber-node.ts';
import {FC} from '#src/disreact/model/entity/fc.ts';



export * as Elem from './element.ts';
export type Elem<P = any, A = any> =
  | RestElement
  | TaskElem
  | LeafElem;

export const Fragment = undefined;

export const make = (type: any, props: any) => {
  switch (typeof type) {
    case 'undefined': {
      return props.children;
    }

    case 'string': {
      return RestElement.make(type, props);
    }

    case 'function': {
      return TaskElem.make(type, props);
    }

    default: {
      throw new Error(`Invalid Element Type: ${type}`);
    }
  }
};

export const clone = <A extends Elem>(self: A): A => {
  if (LeafElem.is(self)) {
    return LeafElem.clone(self) as A;
  }
  if (RestElement.is(self)) {
    return RestElement.clone(self) as A;
  }
  return TaskElem.clone(self) as A;
};

export const deepClone = <A extends Elem>(self: A): A => {
  const cloned = clone(self);

  if (!LeafElem.is(cloned)) {
    for (let i = 0; i < cloned.children.length; i++) {
      cloned.children[i] = deepClone(cloned.children[i]);
    }
  }

  return cloned;
};

export const linearize = <A extends Elem>(self: A): A => {
  if (LeafElem.is(self)) {
    return self;
  }

  if (TaskElem.is(self)) {
    delete self.fiber.root;
    delete self.fiber.element;
  }

  for (let i = 0; i < self.children.length; i++) {
    linearize(self.children[i]);
  }

  return self;
};

export const isSame = (a: Elem, b: Elem): boolean => {
  if (a === b) {
    return true;
  }
  if (a._tag !== b._tag) {
    return false;
  }
  if (a.constructor.name !== b.constructor.name) {
    return false;
  }
  if (LeafElem.is(a)) {
    return a.value === (b as LeafElem).value;
  }
  return true;
};

export interface Meta {
  id     : string;
  idx    : string;
  step_id: string;
}

export const setIds = (children: Elem[], parent: Elem) => {
  for (let i = 0; i < children.length; i++) {
    const child = children[i];

    if (LeafElem.is(child)) {
      continue;
    }
    if (RestElement.is(child)) {
      child.idx = `${child.type}:${i}`;
    }
    if (TaskElem.is(child)) {
      child.idx = `${FC.getName(child.type)}:${i}`;
    }
    child.step_id = `${(parent as TaskElem).idx}:${child.idx}`;
    child.id      = `${(parent as TaskElem).id}:${child.idx}`;
  }
  return children;
};

export const collectStates = (node: Elem, states: { [K in string]: FiberNode } = {}): typeof states => {
  if (TaskElem.is(node)) {
    states[node.id] = node.fiber;
  }

  if ('children' in node && Array.isArray(node.children)) {
    for (const child of node.children) {
      collectStates(child, states);
    }
  }

  return states;
};

export const reduceToStacks = (hooks: { [K in string]: FiberNode }): { [K in string]: FiberNode['stack'] } => {
  return Object.fromEntries(
    Object.entries(hooks)
      .filter(([_, value]) => value.stack.length)
      .map(([key, value]) => [key, value.stack]),
  );
};
