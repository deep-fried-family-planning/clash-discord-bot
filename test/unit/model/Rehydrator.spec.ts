import * as Element from '#src/disreact/model/entity/core/exp/element.ts';
import {Rehydrator} from '#src/disreact/model/Rehydrator.ts';
import {TestDialog} from '#unit/components/test-dialog.tsx';
import {it} from '@effect/vitest';
import * as E from 'effect/Effect';

it.effect('when initialized', E.fn(
  function* () {
    const actual = yield* Rehydrator;

    expect(actual).toMatchInlineSnapshot(`
      Rehydrator {
        "checkout": [Function],
        "register": [Function],
        "rehydrate": [Function],
      }
    `);
  },
  E.provide(Rehydrator.Default()),
));

describe('given no sources', () => {
  it.effect('when registering', E.fn(
    function* () {
      yield* Rehydrator.register(TestDialog);
      const actual = yield* Rehydrator.checkout(TestDialog, {}, {});

      expect(actual).toMatchSnapshot('then the rehydrant envelope is initialized');
    },
    E.provide(Rehydrator.Default()),
  ));

  it.effect('when checking out', E.fn(
    function* () {
      const actual = yield* Rehydrator.checkout(TestDialog, {}, {}).pipe(E.merge);

      expect(actual).toMatchSnapshot('then a defect is returned');
    },
    E.provide(Rehydrator.Default()),
  ));

  it.effect('when rehydrating', E.fn(
    function* () {
      const actual = yield* Rehydrator.checkout(TestDialog, {}, {}).pipe(E.merge);

      expect(actual).toMatchSnapshot('then a defect is returned');
    },
    E.provide(Rehydrator.Default()),
  ));
});

describe('given function component source', () => {
  const layer = Rehydrator.Default({sources: [TestDialog]});

  it.effect('when registering again', E.fn(
    function* () {
      const actual = yield* Rehydrator.register(TestDialog).pipe(E.merge);

      expect(actual).toMatchSnapshot('then a defect is returned');
    },
    E.provide(layer),
  ));

  it.effect('when checking out', E.fn(
    function* () {
      const actual = yield* Rehydrator.checkout(TestDialog, {}, {});

      expect(actual).toMatchSnapshot('then the rehydrant envelope is initialized');
    },
    E.provide(layer),
  ));

  it.effect('when rehydrating', E.fn(
    function* () {
      const actual = yield* Rehydrator.rehydrate(
        {
          id    : TestDialog.name,
          stacks: {},
          props : {},
        },
        {data: 'data'},
      );

      expect(actual).toMatchSnapshot('then the rehydrant envelope is initialized');
    },
    E.provide(layer),
  ));
});

describe('given element source', () => {
  const source = Element.jsx(TestDialog, {});
  const layer = Rehydrator.Default({sources: [source]});

  it.effect('when registering again', E.fn(
    function* () {
      const actual = yield* Rehydrator.register(source).pipe(E.merge);

      expect(actual).toMatchSnapshot('then a defect is returned');
    },
    E.provide(layer),
  ));

  it.effect('when checking out', E.fn(
    function* () {
      const actual = yield* Rehydrator.checkout(source, {}, {});

      expect(actual).toMatchSnapshot('then the rehydrant envelope is initialized');
    },
    E.provide(layer),
  ));

  it.effect('when rehydrating', E.fn(
    function* () {
      const actual = yield* Rehydrator.rehydrate(
        {
          id    : TestDialog.name,
          stacks: {},
          props : {},
        },
        {data: 'data'},
      );

      expect(actual).toMatchSnapshot('then the rehydrant envelope is initialized');
    },
    E.provide(layer),
  ));
});
