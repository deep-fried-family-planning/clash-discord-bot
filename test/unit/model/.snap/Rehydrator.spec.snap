// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given element source > when checking out > then the rehydrant envelope is initialized 1`] = `
{
  "data": {},
  "id": "TestDialog",
  "next": {
    "id": "TestDialog",
  },
  "queue": Set {},
  "root": {
    "$d": 0,
    "$i": 0,
    "$p": 0,
    "_n": "TestDialog:0",
    "_s": "TestDialog:0",
    "name": "TestDialog",
    "pipe": [Function],
    "props": {
      Symbol(disreact/props): Symbol(disreact/props),
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
    "type": [Function],
    Symbol(disreact/source): "TestDialog",
    Symbol(effect/Equal): [Function],
    Symbol(disreact/element): 3,
    Symbol(effect/Hash): [Function],
  },
  "stack": {
    "list": {
      "_id": "MutableList",
      "values": [],
    },
    "seen": WeakSet {},
  },
  "trie": {},
}
`;

exports[`given element source > when registering again > then a defect is returned 1`] = `[SourceDefect: Source (TestDialog) already registered.]`;

exports[`given element source > when rehydrating > then the rehydrant envelope is initialized 1`] = `
{
  "data": {
    "data": "data",
  },
  "id": "TestDialog",
  "next": {
    "id": "TestDialog",
  },
  "queue": Set {},
  "root": {
    "$d": 0,
    "$i": 0,
    "$p": 0,
    "_n": "TestDialog:0",
    "_s": "TestDialog:0",
    "name": "TestDialog",
    "pipe": [Function],
    "props": {
      Symbol(disreact/props): Symbol(disreact/props),
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
    "type": [Function],
    Symbol(disreact/source): "TestDialog",
    Symbol(effect/Equal): [Function],
    Symbol(disreact/element): 3,
    Symbol(effect/Hash): [Function],
  },
  "stack": {
    "list": {
      "_id": "MutableList",
      "values": [],
    },
    "seen": WeakSet {},
  },
  "trie": {},
}
`;

exports[`given function component source > when checking out > then the rehydrant envelope is initialized 1`] = `
{
  "data": {},
  "id": "TestDialog",
  "next": {
    "id": "TestDialog",
  },
  "queue": Set {},
  "root": {
    "$d": 0,
    "$i": 0,
    "$p": 0,
    "_n": "TestDialog:0",
    "_s": "TestDialog:0",
    "name": "TestDialog",
    "pipe": [Function],
    "props": {
      Symbol(disreact/props): Symbol(disreact/props),
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
    "type": [Function],
    Symbol(disreact/source): "TestDialog",
    Symbol(effect/Equal): [Function],
    Symbol(disreact/element): 3,
    Symbol(effect/Hash): [Function],
  },
  "stack": {
    "list": {
      "_id": "MutableList",
      "values": [],
    },
    "seen": WeakSet {},
  },
  "trie": {},
}
`;

exports[`given function component source > when registering again > then a defect is returned 1`] = `[SourceDefect: Source (TestDialog) already registered.]`;

exports[`given function component source > when rehydrating > then the rehydrant envelope is initialized 1`] = `
{
  "data": {
    "data": "data",
  },
  "id": "TestDialog",
  "next": {
    "id": "TestDialog",
  },
  "queue": Set {},
  "root": {
    "$d": 0,
    "$i": 0,
    "$p": 0,
    "_n": "TestDialog:0",
    "_s": "TestDialog:0",
    "name": "TestDialog",
    "pipe": [Function],
    "props": {
      Symbol(disreact/props): Symbol(disreact/props),
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
    "type": [Function],
    Symbol(disreact/source): "TestDialog",
    Symbol(effect/Equal): [Function],
    Symbol(disreact/element): 3,
    Symbol(effect/Hash): [Function],
  },
  "stack": {
    "list": {
      "_id": "MutableList",
      "values": [],
    },
    "seen": WeakSet {},
  },
  "trie": {},
}
`;

exports[`given no sources > when checking out > then a defect is returned 1`] = `[SourceDefect: Source (TestDialog) is not registered.]`;

exports[`given no sources > when registering > then the rehydrant envelope is initialized 1`] = `
{
  "data": {},
  "id": "TestDialog",
  "next": {
    "id": "TestDialog",
  },
  "queue": Set {},
  "root": {
    "$d": 0,
    "$i": 0,
    "$p": 0,
    "_n": "TestDialog:0",
    "_s": "TestDialog:0",
    "name": "TestDialog",
    "pipe": [Function],
    "props": {
      Symbol(disreact/props): Symbol(disreact/props),
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
    "type": [Function],
    Symbol(disreact/source): "TestDialog",
    Symbol(effect/Equal): [Function],
    Symbol(disreact/element): 3,
    Symbol(effect/Hash): [Function],
  },
  "stack": {
    "list": {
      "_id": "MutableList",
      "values": [],
    },
    "seen": WeakSet {},
  },
  "trie": {},
}
`;

exports[`given no sources > when registering and checking out > then the rehydrant envelope is initialized 1`] = `
{
  "data": {},
  "id": "TestDialog",
  "next": {
    "id": "TestDialog",
  },
  "queue": Set {},
  "root": {
    "$d": 0,
    "$i": 0,
    "$p": 0,
    "_n": "TestDialog:0",
    "_s": "TestDialog:0",
    "name": "TestDialog",
    "pipe": [Function],
    "props": {
      Symbol(disreact/props): Symbol(disreact/props),
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
    "type": [Function],
    Symbol(disreact/source): "TestDialog",
    Symbol(effect/Equal): [Function],
    Symbol(disreact/element): 3,
    Symbol(effect/Hash): [Function],
  },
  "stack": {
    "list": {
      "_id": "MutableList",
      "values": [],
    },
    "seen": WeakSet {},
  },
  "trie": {},
}
`;

exports[`given no sources > when rehydrating > then a defect is returned 1`] = `[SourceDefect: Source (TestDialog) is not registered.]`;

exports[`when registering and checking out > then the rehydrant envelope is initialized 1`] = `
{
  "data": {},
  "id": "TestDialog",
  "next": {
    "id": "TestDialog",
  },
  "queue": Set {},
  "root": {
    "$d": 0,
    "$i": 0,
    "$p": 0,
    "_n": "TestDialog:0",
    "_s": "TestDialog:0",
    "name": "TestDialog",
    "pipe": [Function],
    "props": {
      Symbol(disreact/props): Symbol(disreact/props),
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
    "type": [Function],
    Symbol(disreact/source): "TestDialog",
    Symbol(effect/Equal): [Function],
    Symbol(disreact/element): 3,
    Symbol(effect/Hash): [Function],
  },
  "stack": {
    "list": {
      "_id": "MutableList",
      "values": [],
    },
    "seen": WeakSet {},
  },
  "trie": {},
}
`;
